### DISCUSSION TITLE
Struggling with fitz.Pixmap (RuntimeError: unknown image file format)

### DISCUSSION THREAD

**Discussion by axlgit** [pymupdf/PyMuPDF#973] [Q&A] [answered]
Hi,

I've got a PNG image described as numpy ndarray.
I think the ndarray is well shaped because when I passed it to Pytesseract it is accepted and Tesseract gets the text out of it.
Now I would like the same PNG image described as Numpy ndarray to be passed to fitz.Pixmap so that (I guess) it can be inserted into a PDF page by using to .InsertImage and the pixmap input.

To try to understand how to proceed I used this article and the PyMuPDF doc

[https://code.activestate.com/recipes/580630-saving-a-numpy-array-to-png-image-using-fitz-mupdf/](https://code.activestate.com/recipes/580630-saving-a-numpy-array-to-png-image-using-fitz-mupdf/)

[https://documentation.help/PyMuPDF/page.html](https://documentation.help/PyMuPDF/page.html)

but when I run this code I get the error about "image format not recognized".
(The module tools is just a module of mine with a set of procedure to select which images to load from a directory.)

```
>>> import numpy, fitz
>>> from tools import *
>>>
>>> image=select_images('TEST_page0001_trimmed2_rotated_whitened.png')
>>> image = load_images(image)
**>>> image_np_arr = np.ndarray(buffer=image[0].image.write_to_memory(),dtype=np.uint8,shape=[image[0].h,image[0].w,image[0].image.bands])**
>>> out_pdf = fitz.open()
>>> fmt = fitz.PaperRect("a4")
>>> out_page = out_pdf.newPage(width = fmt.width, height = fmt.height)
>>> pix = fitz.Pixmap((fitz.csRGB, image[0].w, image[0].h, image[0].image.bands, bytearray(image_np_arr.tobytes())))
mupdf: unknown image file format
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\axl\AppData\Local\Programs\Python\Python39\lib\site-packages\fitz\fitz.py", line 6389, in __init__
    _fitz.Pixmap_swiginit(self, _fitz.new_Pixmap(*args))
RuntimeError: unknown image file format
```

The assignement to the variable image_np_arr doesn't throw any error (I had just to switch to .tobytes() because the code told me .tostring() was deprecated)

As you can see the image is a PNG and there are 4 bands 

```
>>> image[0].image.bands
4
```

How can I fix this? Or how to troubleshoot  it more?

Thanks in advance,

Alex

**Comment by JorjMcKie** [comment]
You are using the wrong object constructor pattern when defining the pixmap: review the brackets!
You are passing **_one_** argument to the constructor, in which case it must be one of
* a string that defines a filename
* a pathlib.Path object
* a bytes object that contains a valid image file content

Your argument falls in none of these categories.

**Reply by JorjMcKie** [reply]
And because the pixmap constructor correctly excludes the first 2 possibilities, it treats the argument as a bytes object and compares its first dozen or so bytes with the valid image signatures ... and fails of course!

**Reply by axlgit** [reply]
Hi Jori,

sorry for such a basic mistake!
I corrected the number of brackets but I still receive an error.

```
C:\Users\axl\Scripts\>python
Python 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>>
>>> import numpy, fitz
>>> from tools import *
>>>
>>> image=select_images('TEST_page0001_trimmed2_rotated_whitened.png')
>>> image = load_images(image)
>>> image_np_arr = np.ndarray(buffer=image[0].image.write_to_memory(),dtype=np.uint8,shape=[image[0].h,image[0].w,image[0].image.bands])
>>> out_pdf = fitz.open()
>>> fmt = fitz.PaperRect("a4")
>>> out_page = out_pdf.newPage(width = fmt.width, height = fmt.height)
>>>
>>> pix = fitz.Pixmap(fitz.csRGB, image[0].w, image[0].h, bytearray(image_np_arr.tobytes()))
mupdf: bad samples length
Exception ignored in: <function Pixmap.__del__ at 0x000002417FB13820>
Traceback (most recent call last):
  File "C:\Users\axl\AppData\Local\Programs\Python\Python39\lib\site-packages\fitz\fitz.py", line 6691, in __del__
    self.__swig_destroy__(self)
TypeError: in method 'delete_Pixmap', argument 1 of type 'struct Pixmap *'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\axl\AppData\Local\Programs\Python\Python39\lib\site-packages\fitz\fitz.py", line 6389, in __init__
    _fitz.Pixmap_swiginit(self, _fitz.new_Pixmap(*args))
RuntimeError: bad samples length
>>>
>>> fitz.version
('1.18.9', '1.18.0', '20210226134632')
>>>
```

Alex

**Comment by JorjMcKie** [comment, marked as answer]
> bad samples length

This error means ``width * height * (n + alpha)`` does not equal the length of the bytes object, which is used as ``samples`` in the constructor.
The factor ``(n + alpha)`` means: n = 3 for RGB, 1 for GRAY, 4 for CMYK, ``alpha`` is 0 or 1 depending on whether an alpha channel is present.

**Reply by axlgit** [reply]
Thanks Jori,

Thanks for your explanation of error, I'm not very familiar with all this.
Eventually I managed to have fitz accept the Pixmap command in this way:

```
>>> import numpy, fitz
>>> fff = pyvips.Image.new_from_file('TEST_page0001_trimmed2_rotated_whitened.png')
>>> fff.height
1369
>>> fff.width
1038
>>> fff.bands
4
>>> image_np_arr = np.ndarray(buffer=fff.write_to_memory(),dtype=np.uint8,shape=[fff.height,fff.width,fff.bands])
>>> len(bytearray(image_np_arr.tobytes()))
5684088
>>> 1369*1038*4
5684088
>>>


>>> pix = fitz.Pixmap(fitz.csCMYK, fff.width, fff.height, bytearray(image_np_arr.tobytes()))


>>>
```

The picture's info
```
axl@scoobydoo:/BIN$ identify TEST_page0001_trimmed2_rotated_whitened.png
TEST_page0001_trimmed2_rotated_whitened.png PNG 1038x1369 1038x1369+0+0 8-bit sRGB 92.6KB 0.000u 0:00.000
```

I thought that 4 bands would mean RGB + alpha=True but it doesn't seem so.
Is fitz.csCMYK the right way to call Pixmap in my case?

**Reply by JorjMcKie** [reply]
It is a little confusing - but only at first sight:
Both, classes `Pixmap` and `Colorspace` have an attribute named `n`. There relationship in a concrete pixmap is `pix.n = colorspace.n + pix.alpha`. So if you for an RGB image, you **_always_** have `colorspace.n = 3`, but its pix.n may be 3 or 4, depending on whether it is transparent or not.
Similarly for the subtractive colorspace CMYK, which can have `pix.n` 4 or 5.

**Reply by JorjMcKie** [reply]
All questions answered?

**Reply by axlgit** [reply]
Hi Jorj,

sorry for the late reply but I could resume work on this only today.

```
>>> import fitz
>>> import pyvips as pv
>>> import numpy as np
>>>
>>> FILENAME_PNG = 'Screenshot.png'
>>> FILENAME_PDF = 'TEST_PAGE_TEST.pdf'
>>>
>>> out_pdf = fitz.open()
>>> rect = fitz.PaperRect("a4")
>>> out_page = out_pdf.newPage(width = rect.width, height = rect.height)
>>>
>>> a = pv.Image.new_from_file(FILENAME_PNG)
>>> np_arr = np.ndarray(buffer=a.write_to_memory(), dtype=np.uint8,shape=[a.height, a.width, a.bands])
>>> a.hasalpha()
1
>>> a.height
1080
>>> a.width
1920
>>> a.bands
4
>>> a.format
'uchar'
```

I refer to [https://pymupdf.readthedocs.io/en/latest/pixmap.html](https://pymupdf.readthedocs.io/en/latest/pixmap.html)
From the output above I understand my image is standard RGB (colorspace.n = 3) and has the alpha channel, hence in calling fitz.Pixmap I have to pass values as specified by  **__init__(self, colorspace, width, height, samples, alpha)** being colorspace = csRGB and alpha = True but the option alpha is not recognized.

```
>>> pix = fitz.Pixmap(fitz.csRGB, a.height, a.width, bytearray(np_arr.tobytes()), alpha = True)
Exception ignored in: <function Pixmap.__del__ at 0x000002A5A9C961F0>
Traceback (most recent call last):
  File "C:\Users\axl\AppData\Local\Programs\Python\Python39\lib\site-packages\fitz\fitz.py", line 6691, in __del__
    self.__swig_destroy__(self)
TypeError: in method 'delete_Pixmap', argument 1 of type 'struct Pixmap *'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() got an unexpected keyword argument 'alpha'
```

ImageMagick Identify also says the image is sRGB with alpha channel, see attachment
[Screenshot.txt](https://github.com/pymupdf/PyMuPDF/files/6255648/Screenshot.txt)
What I'm doing wrong? And what would be the right way to call Pixmap?

TIA, Alex

**Reply by JorjMcKie** [reply]
``alpha`` is not a keyword! Simply put in 0 (no alpha) or 1.

**Reply by axlgit** [reply]
Great, now it all works, thanks for your patience with me!
